niceman/interface/diff.py

```diff
@@ -0,0 +1,115 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil; coding: utf-8 -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the niceman package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Report if a specification satisfies the requirements in another specification
+"""
+
+import sys
+import time
+
+from .base import Interface
+from ..support.constraints import EnsureStr
+from ..support.exceptions import InsufficientArgumentsError
+from ..support.param import Parameter
+from niceman.formats.niceman import NicemanProvenance
+from ..distributions.debian import DebianDistribution
+from ..distributions.conda import CondaDistribution
+
+__docformat__ = 'restructuredtext'
+
+from logging import getLogger
+lgr = getLogger('niceman.api.retrace')
+
+
+class MultipleDistributionsError(Exception):
+    """Multiple distributions of a given type found"""
+
+    def __init__(self, cls):
+        self.cls = cls
+
+
+class Diff(Interface):
+    """Report if a specification satisfies the requirements in another 
+    specification
+
+    Examples
+    --------
+
+      $ niceman diff --env environment.yml --req required_environment.yml
+
+    """
+
+    _params_ = dict(
+        prov1=Parameter(
+            doc="NICEMAN provenance file", 
+            metavar='prov1',
+            constraints=EnsureStr()),
+        prov2=Parameter(
+            metavar="prov2",
+            doc="NICEMAN provenance file", 
+            constraints=EnsureStr())
+    )
+
+    @staticmethod
+    def __call__(prov1, prov2):
+
+        env_1 = NicemanProvenance(prov1).get_environment()
+        env_2 = NicemanProvenance(prov2).get_environment()
+
+        for (dist_type, pkg_type) in ((DebianDistribution, "Debian package"), 
+                                      (CondaDistribution, "Conda package")):
+
+            dist_1 = env_1.get_distribution(dist_type)
+            pkgs_1 = { p._cmp_id: p for p in dist_1.packages }
```

Comment:
To try to test out this command, I ran `niceman retrace /usr/bin/diff > deb0.yml`.  Then I copied the file to deb1.yml and manually changed the version of the program.  Running `niceman diff deb0.yml deb1.yml` I get the following error

```
'NoneType' object has no attribute 'packages' [diff.py:__call__:68] (AttributeError)
```

In this case, this happens when get_distributions returns None for the conda dist_type.

Reply:
Good catch; thanks.

I fixed this in commit d52ecb7.  The semantics I'm using are that the distributions are nothing more than containers for packages (and are uninteresting in themselves), so the report only shows the differences in contained packages.  Try with:

    niceman diff niceman/interface/tests/files/empty.yaml niceman/interface/tests/files/diff_1.yaml

What do you think?

Reply:
LGTM